exchanges/kucoin/kucoin_websocket.go

```diff
@@ -1678,13 +1724,20 @@ func (o *orderbookManager) stopNeedsFetchingBook(pair currency.Pair, assetType a
 	return nil
 }
 
-func (ku *Kucoin) listOfAssetsCurrencyPairEnabledFor(cp currency.Pair) []asset.Item {
-	assets := []asset.Item{}
-	for _, a := range ku.CurrencyPairs.GetAssetTypes(true) {
-		pairs, err := ku.GetEnabledPairs(a)
-		if err == nil && pairs.Contains(cp, true) {
-			assets = append(assets, a)
+func (ku *Kucoin) listOfAssetsCurrencyPairEnabledFor(cp currency.Pair) ([]asset.Item, error) {
```

Comment:
So..... While I appreciate your improvements here, I'd like to try and address this here if you're cool with that. If not, we can look into it for a future PR, that's okay. This is something I'm working on, I'll look at it with you tomorrow if you want as I expand testing for it.

I would like this function to more easily figure out the asset for the pair without calling such heavy functions.

Here's my code so far:
```go
func (ku *Kucoin) CalculateAssets(topic string, cp currency.Pair) ([]asset.Item, error) {
	if cp.Quote.Equal(currency.USDTM) || strings.Contains(topic, "contract") {
		if err := ku.CurrencyPairs.IsAssetEnabled(asset.Futures); err != nil {
			return nil, nil
		}
		return []asset.Item{asset.Futures}, nil
	}
	if strings.Contains(topic, "margin") {
		if err := ku.CurrencyPairs.IsAssetEnabled(asset.Margin); err != nil {
			return nil, nil
		}
		return []asset.Item{asset.Margin}, nil
	}
	cps := cp.String()
	resp := make([]asset.Item, 0, 2)
	_, spotEnabled, err := ku.MatchSymbolCheckEnabled(cps, asset.Spot, true)
	if err != nil && !errors.Is(currency.ErrCurrencyNotFound, err) {
		return nil, err
	}
	if spotEnabled {
		resp = append(resp, asset.Spot)
	}
	_, marginEnabled, err := ku.MatchSymbolCheckEnabled(cps, asset.Margin, true)
	if err != nil && !errors.Is(currency.ErrCurrencyNotFound, err) {
		return nil, err
	}
	if marginEnabled {
		resp = append(resp, asset.Margin)
	}
	return resp, nil
}


```

Here is my benchmarking:
```go
func BenchmarkButts(b *testing.B) {
	b.ReportAllocs()
	cases := []struct {
		pair           currency.Pair
		expectedResult []asset.Item
	}{
		{pair: spotTradablePair, expectedResult: []asset.Item{asset.Spot}},
		{pair: marginTradablePair, expectedResult: []asset.Item{asset.Spot, asset.Margin}},
		{pair: futuresTradablePair, expectedResult: []asset.Item{asset.Futures}},
	}
	for i := 0; i < b.N; i++ {
		for j := range cases {
			resp, err := ku.listOfAssetsCurrencyPairEnabledFor(cases[j].pair)
			if err != nil {
				b.Fatal(err)
			}
			for x := range resp {
				var found bool
				for y := range cases[j].expectedResult {
					if resp[x] == cases[j].expectedResult[y] {
						found = true
						break
					}
				}
				if !found {
					b.Error("bad match")
				}
			}
		}
	}
}

func BenchmarkButts2(b *testing.B) {
	b.ReportAllocs()
	cases := []struct {
		pair           currency.Pair
		topic          string
		expectedResult []asset.Item
	}{
		{pair: spotTradablePair, topic: "/market/ticker:all", expectedResult: []asset.Item{asset.Spot}},
		{pair: marginTradablePair, topic: "/market/ticker:all", expectedResult: []asset.Item{asset.Margin, asset.Spot}},
		{pair: futuresTradablePair, topic: "/contractMarket/tickerV2:XBTUSDM", expectedResult: []asset.Item{asset.Futures}},
	}
	for i := 0; i < b.N; i++ {
		for j := range cases {
			resp, err := ku.CalculateAssets(cases[j].topic, cases[j].pair)
			if err != nil {
				b.Fatal(err)
			}
			for x := range resp {
				var found bool
				for y := range cases[j].expectedResult {
					if resp[x] == cases[j].expectedResult[y] {
						found = true
						break
					}
				}
				if !found {
					b.Error("bad match")
				}
			}
		}
	}
}
```
here are the results:
```
BenchmarkButts-10     	   16918	     72570 ns/op	  268232 B/op	      25 allocs/op
BenchmarkButts2-10    	 1574653	       763.8 ns/op	      96 B/op	      13 allocs/op
```

Reply:
Updated to account for both private/public margin types and gone over each current use-case and kucoins docs.
Today's version is 6x faster than yesterday's and 420x (😎) faster than current. Hopefully it fits things better for you